src/Hosting/test/FunctionalTests/ShutdownTests.cs

```diff
@@ -0,0 +1,161 @@
+        public async Task ShutdownTestRun()
+        {
+            await ExecuteShutdownTest(nameof(ShutdownTestRun), "Run");
+        }
+
+        [ConditionalFact]
+        [OSSkipCondition(OperatingSystems.Windows | OperatingSystems.MacOSX)]
+        public async Task ShutdownTestWaitForShutdown()
+        {
+            await ExecuteShutdownTest(nameof(ShutdownTestWaitForShutdown), "WaitForShutdown");
+        }
+
+        private async Task ExecuteShutdownTest(string testName, string shutdownMechanic)
+        {
+            using (StartLog(out var loggerFactory))
+            {
+                var logger = loggerFactory.CreateLogger(testName);
+
+                var applicationPath = Path.Combine(TestPathUtilities.GetSolutionRootDirectory("Extensions"),
+                    "src", "Hosting", "test", "testassets", "Microsoft.Extensions.Hosting.TestApp");
+
+                var deploymentParameters = new DeploymentParameters(
+                    applicationPath,
+                    RuntimeFlavor.CoreClr,
+                    RuntimeArchitecture.x64)
+                {
+                    TargetFramework = Tfm.NetCoreApp30,
+                    ApplicationType = ApplicationType.Portable,
+                    PublishApplicationBeforeDeployment = true,
+                    StatusMessagesEnabled = false
+                };
+
+                deploymentParameters.EnvironmentVariables["DOTNET_STARTMECHANIC"] = shutdownMechanic;
+
+                using (var deployer = new SelfHostDeployer(deploymentParameters, loggerFactory))
+                {
+                    var result = await deployer.DeployAsync();
+
+                    var started = new ManualResetEventSlim();
+                    var completed = new ManualResetEventSlim();
+                    var output = string.Empty;
+                    deployer.HostProcess.OutputDataReceived += (sender, args) =>
+                    {
+                        if (!string.IsNullOrEmpty(args.Data) && args.Data.StartsWith(StartedMessage))
+                        {
+                            started.Set();
+                            output += args.Data.Substring(StartedMessage.Length) + '\n';
+                        }
+                        else
+                        {
+                            output += args.Data + '\n';
+                        }
+
+                        if (output.Contains(CompletionMessage))
+                        {
+                            completed.Set();
+                        }
+                    };
+
+                    started.Wait(50000);
+
+                    if (!started.IsSet)
+                    {
+                        throw new InvalidOperationException("Application did not start successfully");
+                    }
+
+                    SendShutdownSignal(deployer.HostProcess);
+
+                    completed.Wait(50000);
+
+                    WaitForExitOrKill(deployer.HostProcess);
```

Comment:
SendShutdownSignal already calls WaitForExitOrKill. Remove it.

Reply:
Nope.

Reply:
The WaitForExitOrKill here is killing the deployer. The WaitForExitOrKill in SendShutdownSignal is killing the "kill" process.

Reply:
Ahh I get it now.